CRM Dynamics integration getLeads UTC date mismatch

woutersf opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 4.x)
* a.b for any bug fixes (e.g. 4.0, 4.1, 4.2)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [yes ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes https://github.com/mautic/mautic/issues/10506


#### Summary

This statement Uses Non UTC dates and creates ZULU (UTC) dates from this in the parameters. (It explicitely cuts of the +02) 
`printf('modifiedon ge %sZ', substr($params['start'], 0, '-6'))`
Zulu time: https://www.timeanddate.com/worldclock/timezone/zulu 

The result is that in the countries with (eg +02) different timezones a set of records is fetched that is incomplete.

There should be something like:

```
$given = new DateTime($params['start']);
$given->setTimezone(new DateTimeZone("UTC"));
$oparams['$filter'] = sprintf('modifiedon ge %sZ', $given->format("Y-m-d\TH:i:s"));
```

#### Description:

@nickveenhof    Thanks for the help on my first Mautic bug find!
Reviewers: Please be gentle: This is my first commit 😁 . 

See ISSUE: https://github.com/mautic/mautic/issues/10506

#### Steps to test this PR:

Also in Bug report https://github.com/mautic/mautic/issues/10506

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10507"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

woutersf committed:
attempt to fix UTC date mismatch

RCheesley commented:
Thanks for the PR @woutersf - please can you check that the email address you signed your commits with is on your GitHub profile (doesn't have to be the default one, just on the profile and confirmed) - that should then placate the bot :)

RCheesley commented:
@woutersf it looks like we have a couple of things to tweak.

Code style wise, you need to use [single rather than double quotes](https://github.com/mautic/mautic/pull/10507/checks?check_run_id=3811911208#step:8:31) in this line 

mautic/plugins/MauticCrmBundle/Integration/DynamicsIntegration.php L31

-                    $startDate->setTimezone(new DateTimeZone("UTC"));
+                    $startDate->setTimezone(new DateTimeZone('UTC'));

PHPSTAN is also throwing an error about a class not being found:

https://github.com/mautic/mautic/pull/10507/checks?check_run_id=3811911164#step:8:85

It relates to:

plugins/MauticCrmBundle/Integration/DynamicsIntegration.php        

L508 Error: Instantiated class MauticPlugin\MauticCrmBundle\Integration\DateTime not found.
L509 Error: Instantiated class MauticPlugin\MauticCrmBundle\Integration\DateTimeZone not found.

If we can get those two addressed we should hopefully get the green light! You also need to sign the Contributors Agreement please: https://www.mautic.org/contributor-agreement.

Thanks!

woutersf committed:
use statements and single quotes

woutersf commented:
- Commits are now signed.
- Contributor agreement emailed signed (that process might be easier than it is now) I quote my partner: "You jump to all these hoops do help them with a commit?" no hard feelings but a form or word download (or just a form) could help?
- Changed quotes 
- Added \DateTime and \DateTimezone use statements.

RCheesley commented:
> * no hard feelings but a form or word download (or just a form) could help?

There is a link, right at the top, to do it online. Did you mean that this form was too complex?
![screenshot-www mautic org-2021 10 08-07_43_45](https://user-images.githubusercontent.com/2930593/136504435-2e6b8012-b7c3-4348-a99c-7de6c7cb76d9.png)

> * Commits are now signed.

Can you double check that the email you are using is listed on your GitHub account? There still seems to be a problem which you can see from [this message](https://github.com/mautic/mautic/pull/10507#issuecomment-938064784) (maybe it was the old historical commit though?)

Thanks for addressing the codestyle and other issue, let's get the tests running and check it out! :)

woutersf commented:
My email address is in the github settings. Not sure what to do next?

RCheesley commented:
@cla-bot check

RCheesley commented:
Great - just needs to get the agreement signed (you can do it online using the link at the top of the page) and then fix the codestyle issue. Then we're good to go! :)

woutersf committed:
code style fix

woutersf commented:
> Great - just needs to get the agreement signed (you can do it online using the link at the top of the page) and then fix the codestyle issue. Then we're good to go! :)

Ah yes, it was in my drafts folder 🤦🏻  - My bad. Email en route...  💌

RCheesley commented:
@cla-bot check

woutersf committed:
Merge branch '4.x' of https://github.com/mautic/mautic into mautic_dynamics_integration_timestamp_mismatch

escopecz commented:
@woutersf please add a unit test to this change. It cannot be merged without it.

nickveenhof committed:
Merge branch '4.x' into mautic_dynamics_integration_timestamp_mismatch

woutersf committed:
Added test for date conversion used in Dynamics filters

woutersf committed:
Added test for date conversion used in Dynamics filters

woutersf committed:
Merge branch 'mautic_dynamics_integration_timestamp_mismatch' of github.com:woutersf/mautic into mautic_dynamics_integration_timestamp_mismatch

woutersf committed:
Removed empty line

woutersf committed:
return type hint

woutersf committed:
comment formatting

woutersf committed:
make it a string

escopecz commented on plugins/MauticCrmBundle/Integration/DynamicsIntegration.php:
```diff
@@ -479,6 +481,21 @@ public function getLeads($params = [], $query = null, &$executed = null, $result
         return $executed;
     }
 
+    /**
+     * Convert to UTC date string for API call.
+     *
+     * @param string $paramDate
+     *
+     * @return string
+     */
+    public function getFilterDateUTC($paramDate)
```
Looking good. Let's use the PHP 7.4 types and save some space :)
```suggestion
    /**
     * Convert to UTC date string for API call.
     */
    public function getFilterDateUTC(string $paramDate): string
```

nickveenhof committed:
Merge branch '4.x' into mautic_dynamics_integration_timestamp_mismatch

escopecz committed:
Using param and return types for the new method

escopecz reviewed (dismissed):
The code change looks good. The test is not optimal as a new public method had to be written for it, but it's way better than no test at all 👍 

_I did not test this, just code reviewed_

kuzmany commented:
Close favor to M5 version https://github.com/mautic/mautic/pull/13209
